token-client/src/main/java/com/sap/cloud/security/client/DefaultSpringHttpClientFactory.java

```diff
@@ -0,0 +1,100 @@
+/**
+ * SPDX-FileCopyrightText: 2018-2023 SAP SE or an SAP affiliate company and Cloud Security Client Java contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package com.sap.cloud.security.client;
+
+import com.sap.cloud.security.config.ClientIdentity;
+import com.sap.cloud.security.mtls.SSLContextFactory;
+import org.apache.hc.client5.http.config.ConnectionConfig;
+import org.apache.hc.client5.http.config.RequestConfig;
+import org.apache.hc.client5.http.impl.classic.CloseableHttpClient;
+import org.apache.hc.client5.http.impl.classic.HttpClients;
+import org.apache.hc.client5.http.impl.io.PoolingHttpClientConnectionManagerBuilder;
+import org.apache.hc.client5.http.io.HttpClientConnectionManager;
+import org.apache.hc.client5.http.ssl.SSLConnectionSocketFactoryBuilder;
+import org.apache.hc.core5.http.ssl.TLS;
+import org.apache.hc.core5.util.Timeout;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.http.client.HttpComponentsClientHttpRequestFactory;
+import org.springframework.web.client.RestTemplate;
+
+import javax.net.ssl.SSLContext;
+import java.io.IOException;
+import java.security.GeneralSecurityException;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class DefaultSpringHttpClientFactory implements SpringHttpClientFactory {
+	private static final Logger LOGGER = LoggerFactory.getLogger(DefaultSpringHttpClientFactory.class);
+	final Set<String> httpClientsCreated = Collections.synchronizedSet(new HashSet<>());
+	final ConcurrentHashMap<String, HttpClientConnectionManager> sslConnectionManagers = new ConcurrentHashMap<>();
+	private static final Timeout DEFAULT_TIMEOUT = Timeout.ofSeconds(5);
+	private static final Timeout DEFAULT_SOCKET_TIMEOUT = Timeout.ofSeconds(5);
+	private static final int MAX_CONNECTIONS_PER_ROUTE = 4; // default is 2
+	private static final int MAX_CONNECTIONS = 20;
+
+	@Override
+	public RestTemplate createRestTemplateClient(ClientIdentity clientIdentity) {
+		HttpComponentsClientHttpRequestFactory requestFactory = new HttpComponentsClientHttpRequestFactory();
+		requestFactory.setHttpClient(createClient(clientIdentity));
+		return new RestTemplate(requestFactory);
+	}
+
+	private CloseableHttpClient createClient(ClientIdentity clientIdentity)
+			throws HttpClientException {
```

Comment:
We used to print a conditional warning when creating a client about not using this default client for production. If you think, it would still make sense, we could add it here to the beginning of the method again?

Reply:
I didn't add them here, because we're going to get rid of this message also in the DefaultHttpClientFactory(in separate PR) as per request from cloud sdk and let end users use the Default impl and document it, so that if they see performance degradation they have to provide their own impl. This will solve the issue when 3 implementation are provided: 1st from us, 2nd from cloud sdk and then 3rd from end user and then end user runs into this error https://github.com/SAP/cloud-security-services-integration-library/blob/main-2.x/token-client/src/main/java/com/sap/cloud/security/client/HttpClientFactory.java#L65